elastic/kibana
# [Alerting] Refactor alerts authorization client

Resolves https://github.com/elastic/kibana/issues/98818

## Summary

Refactoring alerts authorization client to be more generic. This uses the existing alerting feature privilege model to provide `all` and `read` privileges to `rules` and `alerts`. There will be a [followup issue](https://github.com/elastic/kibana/issues/99940) to handle subfeature privilege specification. There should be no change to how the rules RBAC works functionally.

* Updated alerting feature privilege builder to include `entity` in the feature privilege string. Allowed entities are `rule` and `alert`
* Added ability to specify `exemptConsumers` in authorization client to handle the alerting use case where `alerts` was a special consumer id with special handling
* Updated terminology within `AlertsAuthorization` class to reflect updated `rule` terminology
* Updated authorization filter builder to return EsQuery or KQL
* Updated authorization filter to allow field names to be specified
* Exposing `getAlertingAuthorizationWithRequest` through alerting plugin start to be used by other plugins
* Renamed `AlertsAuthorization` to `AlertingAuthorization`
* Changes formatting of recorded audit events
  * Audit event types have changed from using `alerts_` prefix to `alerting_` prefix:
    * `alerts_authorization_failure` is now `alerting_authorization_failure`
    * `alerts_unscoped_authorization_failure` is now `alerting_unscoped_authorization_failure`
    * `alerts_authorization_success` is now `alerting_authorization_success`
  * Audit events have been updated to reflect new alerting terminology
    * `alerts` have been renamed to `rules`, for example the audit event `user Authorized to create a ".index-threshold" alert by "app"` will now read `user Authorized to create a ".index-threshold" rule by "app"`
    * `alert instances` have been renamed to `alert`, for example audit events for `muteInstance` and `unmuteInstance` operations will now reference `muteAlert` and `unmuteAlert`

## To Verify
* Verify that existing roles that give different levels of access to rules for different features still work with this PR
* Verify that new roles created in this PR branch work as expected in terms of giving different levels of access to rules for different features.

### Checklist
